LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DownlinkMessageBuilder.cs

```diff
@@ -202,27 +200,50 @@ internal static class DownlinkMessageBuilder
                 1,
                 loRaDevice.Supports32BitFCnt ? fcntDown : null);
 
+            // following calculation is making sure that ReportedRXDelay is chosen if not default,
+            // otherwise for class C devices a LnsRxDelay of 0 is chosen, for class A devices a value of 1 is chosen
+            var lnsRxDelay = Math.Max((loRaDevice.ClassType is LoRaDeviceClassType.C ? RxDelay0 : RxDelay1).ToSeconds(), loRaDevice.ReportedRXDelay.ToSeconds());
+
             // todo: check the device twin preference if using confirmed or unconfirmed down
-            var downlinkMessage = BuildDownstreamMessage(loRaDevice, request.StationEui, logger, radioMetadata.UpInfo.Xtime, datr, loRaRegion.GetDownstreamRX2DataRate(configuration.Rx2DataRate, loRaDevice.ReportedRX2DataRate, logger, deviceJoinInfo), freq, loRaRegion.GetDownstreamRX2Freq(configuration.Rx2Frequency, logger, deviceJoinInfo), lnsRxDelay, ackLoRaMessage, radioMetadata.UpInfo.AntennaPreference);
+            var downlinkMessage = BuildDownstreamMessage(loRaDevice,
+                                                         request.StationEui,
+                                                         logger,
+                                                         radioMetadata.UpInfo.Xtime,
+                                                         receiveWindow == Constants.ReceiveWindow2 ? null : (datr, freq),
+                                                         rx2,
+                                                         (RxDelay)lnsRxDelay,
```

Comment:
nit: I would move the cast earlier so the variable `lnsRxDelay` is strongly-typed as `RxDelay`.